packages/@n8n/task-runner/src/js-task-runner/built-ins-parser/built-ins-parser.ts

```diff
@@ -0,0 +1,135 @@
+		state: BuiltInsParserState,
+		ancestors: Node[],
+	) => {
+		// $(...)
+		const isDollar = node.callee.type === 'Identifier' && node.callee.name === '$';
+		if (!isDollar) return;
+
+		// $(): This is not valid, ignore
+		if (node.arguments.length === 0) {
+			return;
+		}
+
+		const firstArg = node.arguments[0];
+		if (!isLiteral(firstArg)) {
+			// $(variable): Can't easily determine statically, mark all nodes as needed
+			state.markNeedsAllNodes();
+			return;
+		}
+
+		if (typeof firstArg.value !== 'string') {
+			// $(123): Static value, but not a string --> invalid code --> ignore
+			return;
+		}
+
+		// $("node"): Static value, mark 'nodeName' as needed
+		state.markNodeAsNeeded(firstArg.value);
+
+		// Determine how $("node") is used
+		this.handlePrevNodeCall(node, state, ancestors);
+	};
+
+	private handlePrevNodeCall(_node: CallExpression, state: BuiltInsParserState, ancestors: Node[]) {
+		// $("node").item: In a case like this, the execution
+		// engine will traverse back from current node (i.e. the Code Node) to
+		// the "node" node and use `pairedItem`s to find which item is linked
+		// to the current item. So, we need to mark all nodes as needed.
+		// TODO: We could also mark all the nodes between the current node and
+		// the "node" node as needed, but that would require more complex logic.
+		const directParent = ancestors[ancestors.length - 2];
+		if (isMemberExpression(directParent)) {
+			const accessedProperty = directParent.property;
+
+			if (directParent.computed) {
+				// $("node")["item"]
+				if (isLiteral(accessedProperty)) {
+					if (accessedProperty.value === 'item') {
+						state.markNeedsAllNodes();
+					}
+					// Else: $("node")[123]: Static value, but not 'item' --> ignore
+				}
+				// $("node")[variable]
+				else if (isIdentifier(accessedProperty)) {
+					state.markNeedsAllNodes();
+				}
+			}
+			// $("node").item
+			else if (isIdentifier(accessedProperty) && accessedProperty.name === 'item') {
+				state.markNeedsAllNodes();
+			}
+		} else if (isVariableDeclarator(directParent) || isAssignmentExpression(directParent)) {
+			// const variable = $("node") or variable = $("node"):
+			// In this case we would need to track down all the possible use sites
+			// of 'variable' and determine if `.item` is accessed on it. This is
+			// more complex and skipped for now.
+			// TODO: Optimize for this case
+			state.markNeedsAllNodes();
+		} else {
+			// Something else than the cases above. Mark all nodes as needed as it
+			// could be a dynamic access.
+			state.markNeedsAllNodes();
+		}
+	}
+
+	private visitIdentifier = (node: Identifier, state: BuiltInsParserState) => {
+		if (node.name === '$env') {
+			state.markEnvAsNeeded();
+		} else if (node.name === '$input' || node.name === '$json') {
+			state.markInputAsNeeded();
+		} else if (node.name === '$execution') {
+			state.markExecutionAsNeeded();
+		} else if (node.name === '$prevNode') {
+			state.markPrevNodeAsNeeded();
+		}
+	};
```

Comment:
1. We're marking that something is needed every time we visit its corresponding identifier, even though once is enough. Makes me wonder if we could simplify the code by e.g. collecting a set of all identifiers and marking based on that set.

2. Do we have a fallback so that we request all data in case we find an "unknown" dollar-sign var? i.e. a dollar-sign var added to `n8n-workflow` in future that the runner is not yet set up to handle.

Reply:
1. The marking here is setting a boolean value, which doesn't really matter even if we would do it million times. Collecting the identifiers would mean holding more state and possibly iterating that, which means more memory and loop cycles. I prefer doing it on a single walk
2. That's a good question. Nothing prevents the user from adding a `$myVar` variable in their own code and using that. That would mean we unnecessarily fetch all the data in this case. So I think it's finding a balance how much we optimize and how much we prepare for something that might happen in the future. If we do end up adding a new built-in, I would hope we test it using Code Node and we would catch it there. I added a test to that fails if someone adds a new property to the WorkflowDataProxy